### PR TITLE
KEYCLOAK-3142 - Capture ResourceType that triggers an AdminEvent

### PR DESCRIPTION
**Pull request by thomasdarimont**
Introduced new ResourceType enum for AdminEvents which lists
the current supported ResourceTypes for which AdminEvents
can be fired.

Previously it was difficult for custom EventListeners to figure
out which ResourceType triggered an AdminEvent in order
to handle it appropriately, effectively forcing users to parse
the representation.
Having dedicated resource types as a marker on an AdminEvent helps
to ease custom EventListener code.

### CODE REVIEW

File: server-spi/src/main/java/org/keycloak/events/admin/ResourceType.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+ * limitations under the License.
+ */
+package org.keycloak.events.admin;
+
+/**
+ * Represents Keycloak resource types for which {@link AdminEvent AdminEvent's} can be triggered.
+ *
+ * @author <a href="mailto:thomas.darimont@gmail.com">Thomas Darimont</a>
+ */
+public enum ResourceType {
```
Comment (thomasdarimont) [inline server-spi/src/main/java/org/keycloak/events/admin/ResourceType.java]: There is already a similar type in {{org.keycloak.partialimport.ResourceType}} but it is in a seperate package that I didn't want to pull-in. Perhaphs one could unify both enums.

**Comment by stianst** [comment]
@mposolda can you review this please?

**Comment by mposolda** [comment]
PR looks good to me and will be nice to have ResourceType on admin events. Thanks for the contribution Thomas! 

Just few things:
- Liquibase changelog will need to be updated as "2.0.0" is already released. Sorry the PR was in the queue for longer time :( Also is it possible to remove the precondition for existing column? I suppose you added that because you manually change your DB schema, however each precondition adds additional complexity into changelog and may slow down migration a bit. So if it's possible, can you manually remove the column from your DB before running the migration (or rename column and then add data back to new column after migration is finished if you don't want to loose the info in your DB? I am sure you will figure some way to handle this :) )
- Tests will need to be added. We have AssertAdminEvents to test admin events and it's used on many places. I can see the alternatives like:
1) change all places where AssertAdminEvents are called and add resourceType to them
2) create overriden method on AssertAdminEvents, which won't use resourceType, so you won't need to change all the tests.
The option (1) is more work, but IMO will be better as it will test correct resourceType is on all the places.

**Comment by thomasdarimont** [comment]
Hi @mposolda, 

just a quick heads up - I'm applied your suggestions to my local branch and and I'm half-way through with Option 1). I'll try to get it done and pushed this weekend.

Cheers,
Thomas

**Comment by mposolda** [comment]
Thanks for the update @thomasdarimont !

**Comment by thomasdarimont** [comment]
@mposolda the requested test updates are in place now. I just rebased the branch.
Should be good now.

Cheers,
Thomas

**Comment by mposolda** [comment]
@thomasdarimont Thanks for the update. I will followup tomorrow with your PR and will rebase and possibly remove conflicts to have it finally in. However it will be in 2.2.0.CR1 release. Sorry for the delay...

**Comment by thomasdarimont** [comment]
Thanks, no worries :-)

**Comment by mposolda** [comment]
@thomasdarimont I've tried to followup on your PR, so cherry-picked your commit on top of latest master and sent another PR https://github.com/keycloak/keycloak/pull/3112/commits/f407658e605061a9fb303680b12b58739bb83fa6 . The only thing I changed was to move your stuff to new jpa-changelog-2.2.0 and added that to master changelog. 

However I can see lots of test failures in travis. See https://api.travis-ci.org/jobs/149666747/log.txt?deansi=true . Indeed when I tried to run some tests locally, I figured they are failing.

Will you have time to follow and fix the tests? I know that in your PR, there is easy possibility of conflict. So I can follow if there is few tests failing. However not if almost whole testsuite seems to be failing ;)

**Comment by thomasdarimont** [comment]
Thanks @mposolda for having a look - strange, worked for me locally.
How about me rebasing the PR on current master, and moving the changelog stuff to the new 2.2.0 file?
Then I'll have a look at the tests.

Sounds like a plan?

**Comment by mposolda** [comment]
@thomasdarimont Sure, you can do either that or just cherry-pick my commit as that's exactly what I did in my commit.

**Comment by thomasdarimont** [comment]
Ah I found the reason for the test failures - I didn't implement the asserts for the adminEvents correctly for the client tests. I'll push a the changes as a new commit instead of amending the existing that you can see what I fixed.

File: services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java [hunk 2]
```diff
@@ -172,7 +172,6 @@ public void deleteRole(final @PathParam("role-name") String roleName) {
             throw new NotFoundException("Could not find client");
         }
 
-        RoleRepresentation rep = getRole(roleName);
```
Comment (thomasdarimont) [inline services/src/main/java/org/keycloak/services/resources/admin/RoleContainerResource.java]: was not used ... and didn't seem to have a relevant side effect so I removed it.

**Comment by thomasdarimont** [comment]
PR should be good now - somehow the travis build kicked in again? ~13min before.

**Comment by mposolda** [comment]
yep, I ran it again manually :)

**Comment by mposolda** [comment]
@thomasdarimont Thanks for your contribution

**Pull request merged**